### PR TITLE
Alternative network defs

### PR DESCRIPTION
**Pull request by dancrumb**

### CODE REVIEW

File: core/Network.js [hunk 1]
```diff
@@ -5,10 +5,11 @@ var IIPConnection = require('./IIPConnection')
   , Process = require('./Process')
   , ProcessConnection = require('./ProcessConnection')
   , parseFBP = require('parsefbp')
-  , trace = require('./trace');
+  , trace = require('./trace')
+  , _ = require('lodash');
 
 var Network = module.exports = function () {
-  this._processes = [];
```
Comment (dancrumb) [inline core/Network.js]: This is what was killing you (in part).

JS uses `{ }` for objects and associative arrays. It uses `[ ]` for arrays.

When you use `foo['bar'] = 42` on a basic array, it's not strictly an error, but that value doesn't get put _into_ the array, so you get whacky behaviour.

File: core/Network.js [hunk 2]
```diff
@@ -82,11 +81,11 @@ Network.prototype.run = function (runtime, options, callback) {
     port[1].setRuntime(runtime);
   }
 
-  this._processes.forEach(function (process) {
+  _.forEach(this._processes, function (process) {
```
Comment (dancrumb) [inline core/Network.js]: This is the cleanest way to iterate over the _values_ in an assoc. array.

File: core/Network.js [hunk 3]
```diff
@@ -82,11 +81,11 @@ Network.prototype.run = function (runtime, options, callback) {
...
     process.inports.forEach(setPortRuntime);
     process.outports.forEach(setPortRuntime);
   });
-  runtime.run(this._processes, options, callback || function () {
+  runtime.run(_.values(this._processes), options, callback || function () {
```
Comment (dancrumb) [inline core/Network.js]: `runtime.run` takes a simple array of processes, so we need to extract them from the assoc array. This does that.

File: core/Network.js [hunk 4]
```diff
@@ -95,12 +94,24 @@ Network.prototype.defProc = function (func, name) {
     func = loadComponent(func);
   }
   if (!func) {
-    throw new Error("No function passed to defProc");
+    throw new Error("No function passed to defProc: " + name);
+  }
+  if (!name) {
+    name = func.name;
```
Comment (dancrumb) [inline core/Network.js]: Since functions can be anonymous, we need to check for that.

**Comment by jpaulm** [comment]
Added lodash to Readme  We're up to 6 packages now - can this be automated?

Regards,

Paul

**Comment by dancrumb** [comment]
The problem with fbptest05 was a simple quote character mismatch... " then '

**Comment by jpaulm** [comment]
Thanks, well spotted!

**Comment by jpaulm** [comment]
fbptest02.js is now working!  Didja do something (to reader.js) or is it
just a phase of the moon thing?!  Great news... if it holds - I was really
out of my depth here!

Thanks, as always,

Paul

**Comment by dancrumb** [comment]
I think it was just that hinkiness with `Network.js` with the wrong kind of array initialization at line 11 (commented above). That's what fixed the remaining tests!

**Comment by jpaulm** [comment]
I think I get it!  Thanks for explaining!  I'll be more careful in the future!

**Comment by jpaulm** [comment]
Github says the merge worked.  When I did a git pull, I got this:

C:\Users\Paul\Documents\GitHub\jsfbp [alternative_network_defs ↕ +0 ~2 -0
!]> gi
t pull
remote: Counting objects: 28, done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 28 (delta 9), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (28/28), done.
From https://github.com/jpaulm/jsfbp
 + b74c7c4...73fe4dd feature/clarify-ports -> origin/feature/clarify-ports
(for
ced update)
   b86338e..b69b137  master     -> origin/master
Your configuration specifies to merge with the ref
'refs/heads/alternative_netwo
rk_defs'
from the remote, but no such ref was fetched.
C:\Users\Paul\Documents\GitHub\jsfbp [alternative_network_defs ↕ +0 ~2 -0
!]>

Can you explain?  Do I have to do something?!

TIA

**Comment by jpaulm** [comment]
Ouch!  I had deleted alternative_network_defs - now I have two trivial
fixes, and when I did a commit Github recreated the branch.  I have gone
ahead and create a PR - should I bull ahead with the merge, or is there a
way to back out my last commit?

TIA,

Paul

**Pull request merged**